src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/VehicleSharingGraphWriterRunnable.java

```diff
@@ -1,32 +1,77 @@
 package org.opentripplanner.updater.vehicle_sharing.vehicles_positions;
 
-import org.apache.commons.lang3.tuple.Pair;
+import org.opentripplanner.graph_builder.linking.TemporaryStreetSplitter;
 import org.opentripplanner.routing.core.vehicle_sharing.VehicleDescription;
-import org.opentripplanner.routing.edgetype.rentedgetype.RentVehicleAnywhereEdge;
 import org.opentripplanner.routing.graph.Graph;
+import org.opentripplanner.routing.graph.Vertex;
+import org.opentripplanner.routing.vertextype.TemporaryRentVehicleVertex;
+import org.opentripplanner.routing.vertextype.TemporaryVertex;
 import org.opentripplanner.updater.GraphWriterRunnable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import java.util.Collection;
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
 
 class VehicleSharingGraphWriterRunnable implements GraphWriterRunnable {
 
-    private final List<Pair<RentVehicleAnywhereEdge, VehicleDescription>> appeared;
-    private final List<RentVehicleAnywhereEdge> rememberedVehicles;
+    private static final Logger LOG = LoggerFactory.getLogger(VehicleSharingGraphWriterRunnable.class);
+
+    private final TemporaryStreetSplitter temporaryStreetSplitter;
+
+    private final List<VehicleDescription> vehiclesFetchedFromApi;
+
 
-    public VehicleSharingGraphWriterRunnable(List<Pair<RentVehicleAnywhereEdge, VehicleDescription>> appeared,
-                                             List<RentVehicleAnywhereEdge> rememberedVehicles) {
-        this.appeared = appeared;
-        this.rememberedVehicles = rememberedVehicles;
+    VehicleSharingGraphWriterRunnable(TemporaryStreetSplitter temporaryStreetSplitter,
+                                      List<VehicleDescription> vehiclesFetchedFromApi) {
+        this.temporaryStreetSplitter = temporaryStreetSplitter;
+        this.vehiclesFetchedFromApi = vehiclesFetchedFromApi;
     }
 
     @Override
     public void run(Graph graph) {
-        for (RentVehicleAnywhereEdge edge : rememberedVehicles) {
-            // TODO remove only vehicles which disappeared
-            edge.getAvailableVehicles().clear();
-        }
-        for (Pair<RentVehicleAnywhereEdge, VehicleDescription> p : appeared) {
-            p.getLeft().getAvailableVehicles().add(p.getRight());
-        }
+        removeDisappearedRentableVehicles(graph);
+        addAppearedRentableVehicles(graph);
+    }
+
+    private void removeDisappearedRentableVehicles(Graph graph) {
+        Map<VehicleDescription, Optional<TemporaryRentVehicleVertex>> disappearedVehicles = getDisappearedVehicles(graph);
+        List<Vertex> properlyLinkedVertices = getProperlyLinkedVertices(disappearedVehicles.values());
```

Comment:
Rozumiem czym jest zmienna ale nie rozumiem co ma jej nazwa oznaczać

Reply:
jeśli udało się połączyć tymczasowy wierzchołek trzymający dany pojazd na mapę, to jest `properly linked`. Tych, których nie udało się połączyć, nie zapisuję, bo nie są potrzebne. Potrzebne są za to informacje, że nie udało się danego pojazdu rzutować na graf - żeby nie próbować ponownie (stąd możliwość wystąpienia `Optional.empty()`).